### DISCUSSION TITLE
What version should we use in `package.json`?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#476] [General]
The version is currently set to `1.0.0`, but I'm not sure this is good. Might it be better to have the Node.js version match the protocol release (i.e. `2.0.0`)?

https://github.com/sablierhq/v2-core/blob/11a57870cf603895b5d70a950fead68651d35e2b/package.json#L5

**Comment by PaulRBerg** [comment]
I would love to hear your thoughts on this, @razgraf, @gavriliumircea, since the client app will be the leading consumer of these packages.

**Reply by razgraf** [reply]
The package is already called `@sablier/v2-core` so the protocol-level version control is accounted for.

Versions in package.json and other similar configuration files should track the published instances of a specific library. For this **new package** I believe we need to start fresh with a library-level version of 1. 

TL;DR I would version it `1.0.0`

---

For what it's worth, uniswap uses the same heuristic for their [v3-core](https://github.com/Uniswap/v3-core/blob/main/package.json). Same with yield's [vault-v2](https://github.com/yieldprotocol/vault-v2/blob/master/package.json).

**Reply by PaulRBerg** [reply]
> called `@sablier/v2-core` so the protocol-level version control is accounted for

Is it? The name of the package includes only a MAJOR version `v2` - it doesn't tell anything about the MINOR or the PATCH versions.

When we ship v2.1, it will be awkward for the Node.js package to be versioned something like `1.1.0`.

> new package

New package that is inherently connected to the contract ABIs of a specific protocol release.

---

My thinking is to mirror the protocol version in the MAJOR and the MINOR versions (`a` and `b` in `a.b.c`) but reserve the PATCH versions (`c`) for Node.js-related updates only. Protocol deployments will not be shipped often, but Node.js packages may, e.g. non-Solidity-related fixes, or including a previously-unincluded contract artifact in the npm package.

With this approach, the first version of `@sablier/v2-core` would be `2.0.0`. Potential smaller updates would bump the version to `2.0.1`, `2.0.2`, and so forth. Then, if and when we ship Sablier V2.1 (the protocol), we would keep the npm package in sync by bumping it to `v2.1.0`.

**Reply by PaulRBerg** [reply]
Does this make sense to you, @razgraf? @gavriliumircea?

I'm looking to wrap up the development on the Node.js package soon, ref #477.

Now (as opposed to when I created this discussion), I'm pretty staunchly in favor of using `2.0.0` as per the reasons given above.

**Reply by gavriliumircea** [reply]
I do agree with @razgraf(I see v2-core a complete separate library by itself and not an upgraded version of a different protocol), I would use the version 1.b.c and yes we could use the c part just for Node.js-related updates.

**Reply by PaulRBerg** [reply]
Thanks, @gavriliumircea.

> I see v2-core a complete separate library by itself and not an upgraded version of a different protocol

I don't see why we should view `@sablier/v2-core` (or why it is, regardless of our beliefs) as a separate library?

The contract ABIs shipped in this npm package will correspond to a specific Sablier protocol release, which is V2.

With my suggested approach, we would never change the MAJOR version of this package until we ship Sablier V3, which I think it's a sensible approach because it avoids potentially confusing situations like:

1. How should we name the package when V2.1 is shipped? `1.1.0`?
1. If we ship a new MAJOR version after V2.1 is shipped, it would be misleading for the package version to be `2.0.0`

We would have to maintain a mapping on our docs website between protocol releases and package versions. In contrast, if we keep the package version in sync with the protocol release (except for the MINOR versions), it's all nice and smooth.

**Reply by razgraf** [reply]
1.1.0 and I think it's safe to leave "major" versions out of scope as a major version would imply a V3 (separate package). 

I see a versioning system as an incremental ordering rule, following a historically accurate individual timeline. Objectively v2-core is a different package than v1-core. Subjectively, I believe a package should start from version 0 (beta) or version 1 (alpha), where anything after alpha should be traceable back to a `v1.0.0`. If we start from `v2.0.0` you need to "find context" for why there's no v1 for this package/module.

**Reply by razgraf** [reply]
As ChatGPT suggests, it's possible to start from 2.0.0 but "may confuse users and make it difficult to understand the history of the package".

![Screenshot 2023-05-12 at 14 54 56](https://github.com/sablierhq/v2-core/assets/15740806/1759676f-a266-4a55-9a30-13fc94726021)

**Reply by PaulRBerg** [reply]
> 1.1.0

We will have to map Sablier v2.1 to `@sablier/v2-core@1.1.0` on our docs website.

> it's safe to leave "major" versions out of scope as a major version would imply a V3 (separate package).

Never shipping major versions is in accord with my approach to start from `2.0.0`.

It thus follows that our only disagreement is over the major version - whether it should be `1` or `2`.

> I see a versioning system as an incremental ordering rule

In the general case, yes, I see Node.js package similarly. But there's a lot of baggage and path dependence in Sablier, which we can't shrug off as irrelevant.

> a package should start from version 0 (beta) or version 1 (alpha)

Alpha (A) versions come before beta (B) versions, so this should be the other way around.

> If we start from v2.0.0 you need to "find context"

The name of the package is the context.

> As ChatGPT suggests

That was a vague prompt - if we want ChatGPT to provide us with a good answer, we have to explain the background story with Sablier V2, mentioning that the package name contains a major version. In the general case, I agree with ChatGPT that it's best to start from `1.0.0`.

Here's a better prompt:

|    Prompt    |     Response    |
| :----------: | :----------: |
|  ![Before](https://i.postimg.cc/65f9v6Gp/prompt.png)  |  ![After](https://i.postimg.cc/RZhjZv78/response.png)  |

**Reply by razgraf** [reply]
The vague prompt was intentional - I wanted an objective, non-biased answer.

Re. alpha <> beta, my bad, the other way around.

**Reply by PaulRBerg** [reply]
> The vague prompt was intentional - I wanted an objective, non-biased answer.

What is biased and subjective in my prompt? I presented both solutions in an objective way to ChatGPT. I said we are "undecided".

Perhaps you meant to say that you wanted an answer for the general case - which, again, I agree with both you and ChatGPT that it's good to start from `1.0.0` when starting afresh, but that is not applicable to us given the existence of Sablier's history and the inclusion of a numerical version in the name of the release "Sablier V2" (which is a matter of fact, not of opinion).

**Reply by PaulRBerg** [reply]
I just realized another benefit of using `2.0.0` instead of `1.0.0` - keeping git tags and GitHub releases in sync with the release name.

It would be *really* awkward for the version tag and the release on GitHub to have a big title that reads "v1.0.0".

To this, you might retort by saying that we don't need a GitHub release. And yes, maybe we don't need one (I haven't made up my mind about this), but we definitely need a git tag for sharing URLs on social media.

**Reply by razgraf** [reply]
> really awkward

Could you elaborate, why would it be awkward?

> ChatGPT prompt

I'm very intentional with keeping prompts general, as you pointed out and objective with ChatGPT as it is very easy to influence. I ran your prompt through OCR and modified two things in the text
- switched the order of the arguments "one one hand / on the other hand"
- added a line at the end that includes the name of our previous package

I am however using ChatGPT-3. Would be interested to see if 4 keeps its recommendation with slight changes to the prompt.

| Prompt | Response |
|--------|--------|
| <img width="919" alt="Screenshot 2023-05-12 at 16 58 32" src="https://github.com/sablierhq/v2-core/assets/15740806/a11c1a6c-48eb-4a75-85d0-622f1f5642cb"> | <img width="831" alt="Screenshot 2023-05-12 at 16 59 53" src="https://github.com/sablierhq/v2-core/assets/15740806/50931752-e3d9-4090-b00f-87891411bb7e"> |

---

At the end of the day you asked for our opinion so the tl;dr; is that I find v1.0.0 more intuitive as a programmer, while v2.0.0 may feel more intuitive as a general observer. You are of course encouraged to choose whatever makes most sense to move this forward faster. But in terms of **thoughts**, I think v1 > v2 in this case.

**Reply by PaulRBerg** [reply]
> why would it be awkward?

Awkwardness is a second-order effect of all the reasons I have given in my previous replies.

In practice: 

1. We tweet about Sablier v2.0
2. We link to the source code on GitHub
3. Wild "v1.0.0" tag appears as a URL preview

> ChatGPT as it is very easy to influence

First, about the prompt. It's self-contradicting in the last paragraph (see explanation below), and I suspect this has confused ChatGPT.

Second, about the response. Even if the prompt had been logically consistent, we should use ChatGPT to buttress our position(s), not take what it says on faith.

One of the points it made in your prompt's response is "If you make a backward-incompatible change, you would release it as '2.0.0'", which runs counter to what you said above, i.e. "safe to leave major versions out of scope as a major version would imply a V3 (separate package)".

> added a line at the end that includes the name of our previous package

Adding that line has made the last paragraph self-contradictory. There either existed a package or not. Nothing in-between.

What you perhaps intended to say is that while there has never been a published package `@sablier/v1-core`, we had a private package `sablier`. However, this is a red herring because the very topic we are discussing is publishing packages.

> I am however, using ChatGPT-3

ChatGPT4 is a chad compared to ChatGPT3. I highly recommend you get the premium plan if you don't have it already.

> Would be interested to see if 4 keeps its recommendation with slight changes to the prompt.

It does keep the recommendation after changing the order of the OT1H/ OTOH arguments:

|    Prompt    |     Response    |
| :----------: | :----------: |
|  ![Before](https://i.postimg.cc/wMNMHb87/prompt-2.png)  |  ![After](https://i.postimg.cc/C1kpKy5T/response-2.png) |

> you asked for our opinion

As I always do; progress is made through disagreement and debate.

> I find v1.0.0 more intuitive as a programmer, while v2.0.0 may feel more intuitive as a general observer.

I believe v2.0.0 is more intuitive for (some) programmers, too, because of the git tag problem.

**Reply by PaulRBerg** [reply]
Given the above, I will consider the conjecture "v1.0.0 is better than v2.0.0" refuted, and proceed with changing the package version to `v2.0.0` in #477 so that we keep it consistent with the release name, git tags, and so forth.

As ever, I would be happy to change my mind if you guys manage to refute the arguments in my previous comment.

**Comment by PaulRBerg** [comment]
I am coming full circle on this versioning issue.

Working on [V2 Periphery](https://github.com/sablier-labs/v2-periphery) has made me realize that my proposal to version the V2 Core package as `2.0.0` doesn't scale well, albeit not for the reasons discussed above.

The refutation is thus: we can keep a single package in sync with the protocol release name ("V2"), but we cannot keep multiple packages in sync using this approach. Versioning both V2 Core as `2.0.0` and V2 Periphery as `2.0.0` might work initially, but the packages will differentiate over time - what happens, then, when we want to update the `ProxyTarget`? The package versions will have to diverge, i.e., bump V2 Periphery to `2.1.0` or `2.0.1`. Doing so runs counter to my approach of keeping the package versions and the protocol releases in sync. (note: there's a separate issue about whether a proxy target update should be called Sablier V2.1; I think not, but we can talk about this later)

Additionally, there isn't just the V2 Periphery package - there's also [V2 Docs](https://github.com/sablier-labs/v2%20docs), and who knows what other V2-related packages we will ship in the future.

Therefore, I take back my proposal and will move back the package version to `1.0.0`. We will address the git tag problem (which is still a valid problem) by not linking to the git tags directly; we will link only to the `v2-core` home page on GitHub.
